### PR TITLE
Add support for vectors of MultiMooseEnums

### LINKED ISSUE #28487: Support for vectors of MultiMooseEnums
### Motivation

In an Action that sets up multiple versions of an object, we need to parse a vector of MultiMooseEnums.

### Design

Some additions in Builder.C should be all that is required.

### Impact

Add something new

### PR DESCRIPTION
**Pull request by jasondhales**
Closes #28487

### CODE REVIEW

File: framework/src/parser/Builder.C [hunk 1]
```diff
@@ -2162,6 +2171,40 @@ Builder::setVectorParameter<MooseEnum, MooseEnum>(
...
+
+  // Get the full string assigned to the variable full_name
+  std::string buffer = root()->param<std::string>(full_name);
+
+  std::vector<std::string> first_tokenized_vector;
+  MooseUtils::tokenize<std::string>(buffer, first_tokenized_vector, 1, ";");
+  param->set().resize(first_tokenized_vector.size(), enum_values[0]);
+
+  std::vector<std::vector<std::string>> vecvec(first_tokenized_vector.size());
+  for (unsigned j = 0; j < vecvec.size(); ++j)
```
Comment (GiudGiud) [inline framework/src/parser/Builder.C:2193]:
```suggestion
  for (const auto j : index_range(vecvec))
```

File: framework/src/parser/Builder.C [hunk 2]
```diff
@@ -2162,6 +2171,40 @@ Builder::setVectorParameter<MooseEnum, MooseEnum>(
...
+  {
+    MooseUtils::tokenize<std::string>(first_tokenized_vector[j], vecvec[j], 1, " ");
+    param->set()[j] = vecvec[j];
+  }
+
+  if (in_global)
+  {
+    global_block->remove(short_name);
+    global_block->setVectorParam<MultiMooseEnum>(short_name).resize(vecvec.size(), enum_values[0]);
+    for (unsigned int i = 0; i < vecvec.size(); ++i)
```
Comment (GiudGiud) [inline framework/src/parser/Builder.C:2203]:
```suggestion
    for (const auto i : index_range(vecvec))
```

**Review by GiudGiud — COMMENTED** [review]
you'll need to add a test, under test/tests/parser

File: framework/src/utils/InputParameters.C [hunk 3]
```diff
@@ -1143,6 +1143,20 @@ InputParameters::addRequiredParam<std::vector<MooseEnum>>(
...
 
+template <>
+void
+InputParameters::addRequiredParam<std::vector<MultiMooseEnum>>(
+    const std::string & name,
+    const std::vector<MultiMooseEnum> & moose_enums,
+    const std::string & doc_string)
+{
+  InputParameters::set<std::vector<MultiMooseEnum>>(name) =
+      moose_enums; // valid parameter is set by set_attributes
```
Comment (GiudGiud) [inline framework/src/utils/InputParameters.C:1154]: I think we should check (assert) that moose_enums is
- not empty
- size 1 (only one enum is supported later in the parsing, we dont support a parameter like {MultiMooseEnum(a b c), MultiMooseEnum(d e)}

Comment (jasondhales) [reply framework/src/utils/InputParameters.C:1161]: I've added checks for those

**Review by GiudGiud — DISMISSED** [review]

**Comment by GiudGiud** [comment]
I'm worried that test failure is related. We'll see when next is fixed

**Comment by GiudGiud** [comment]
I think some modification of hit or the hit test might be necessary to support this.
The parameter dump is no longer parsed properly based on the failing test

**Comment by jasondhales** [comment]
> I think some modification of hit or the hit test might be necessary to support this. The parameter dump is no longer parsed properly based on the failing test

Any idea on where to start looking?  It must have something to do with the default and required value for the mme parameter in the new test class, but I don't know what's wrong with the dump or how to fix it.

**Comment by GiudGiud** [comment]
I would try to require / ban the default, see if we can get it work with that limitation.

If that does not work, gets more involved. We need to understand how the parameters are getting parsed

**Comment by GiudGiud** [comment]
This isnt working as expected.
```
    mme = 'b; c a;; a b c d e'
```
and 
```
    mme = 'b; c a; a b c d e; '
```
both triger the wrong error

File: test/src/materials/VecMultiMooseEnumMaterial.C [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  InputParameters params = Material::validParams();
+  std::vector<MultiMooseEnum> mme = {MultiMooseEnum("a b c d e")};
+  params.addRequiredParam<std::vector<MultiMooseEnum>>("mme", mme, "Vector of MultiMooseEnums");
+  return params;
+}
+
+VecMultiMooseEnumMaterial::VecMultiMooseEnumMaterial(const InputParameters & parameters)
+  : Material(parameters)
+{
+}
```
Comment (GiudGiud) [inline test/src/materials/VecMultiMooseEnumMaterial.C:26]: let's print the vector parameters here

File: test/tests/parser/vector_multimooseenum/tests [hunk 5]
```diff
@@ -0,0 +1,11 @@
+[Tests]
+  issues = '#28487'
+  design = 'InputParameters.md'
+
+  [vmme]
+    type = 'RunApp'
+    input = 'vmme.i'
+    requirement = 'The system shall support vectors of MultiMooseEnums as input parameters.'
+    expect_out = 'Solve Skipped!'
```
Comment (GiudGiud) [inline test/tests/parser/vector_multimooseenum/tests:9]: and use the print here for the expect out

File: test/tests/parser/vector_multimooseenum/tests [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+  []
```
Comment (GiudGiud) [inline test/tests/parser/vector_multimooseenum/tests:10]: let's add two exception tests: one to make sure the empty vector is caught properly ('a;; b c', and one for the wrong value (like 'z' here) passed in the enum

**Review by GiudGiud — CHANGES_REQUESTED** [review]

File: test/tests/parser/vector_multimooseenum/tests [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    input = 'vmme.i'
+    requirement = 'The system shall support vectors of MultiMooseEnums as input parameters.'
+    expect_out = 'mme: b; c a; a b c d e; e;'
+  []
+  [vmme_empty]
+    type = RunException
+    input = 'vmme.i'
+    cli_args = "Materials/vmme/mme='b;; a b c d e; e'"
+    requirement = 'The system shall return a usage message when the vector of MultiMooseEnums includes an empty entry.'
+    expect_err = "missing required parameter 'Materials/vmme/mme'"
```
Comment (GiudGiud) [inline test/tests/parser/vector_multimooseenum/tests:16]: I must have commented on the commit, I dont see my comment anymore

this isnt the right message, it should say something explicit like "one entry in the vector is empty and this is not allowed"
let's create a feature request issue if you're running out of time on this

Comment (jasondhales) [reply test/tests/parser/vector_multimooseenum/tests]: I gave it another try.  Think this now does what you requested.

**Review by GiudGiud — APPROVED** [review]

File: test/tests/parser/vector_multimooseenum/tests (hunk 7 above)
Comment (GiudGiud) [reply test/tests/parser/vector_multimooseenum/tests]: awesome. thanks

**Comment by loganharbour** [comment]
Reminder to not push directly to idaholab for PRs. This was pushed to idaholab and not a fork.

**Pull request merged**